cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java

```diff
@@ -0,0 +1,220 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.data.batch;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.data.format.FormatSpecification;
+import co.cask.cdap.api.data.stream.StreamBatchReadable;
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import co.cask.cdap.api.stream.StreamEventDecoder;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Defines input to a program, such as MapReduce.
+ */
+public abstract class Input {
+
+  private final String name;
+
+  private Input(String name) {
+    this.name = name;
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   */
+  public static Input ofDataset(String datasetName) {
+    return ofDataset(datasetName, Collections.<String, String>emptyMap());
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param arguments the arguments to use when instantiating the dataset
+   */
+  public static Input ofDataset(String datasetName, Map<String, String> arguments) {
+    return ofDataset(datasetName, arguments, null);
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param splits the data selection splits
+   */
+  public static Input ofDataset(String datasetName, Iterable<Split> splits) {
+    return ofDataset(datasetName, Collections.<String, String>emptyMap(), splits);
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param arguments the arguments to use when instantiating the dataset
+   * @param splits the data selection splits
+   */
+  public static Input ofDataset(String datasetName, Map<String, String> arguments, Iterable<Split> splits) {
+    return new DatasetInput(datasetName, arguments, splits);
+  }
+
+  /**
+   * Returns an Input defined by an InputFormatProvider.
+   *
+   * @param inputName the name of the input
+   */
+  public static Input of(String inputName, InputFormatProvider inputFormatProvider) {
+    return new InputFormatProviderInput(inputName, inputFormatProvider);
+  }
+
+  /**
+   * Returns an Input defined by a stream.
+   *
+   * @param streamBatchReadable specifies the stream to be used as input
+   */
+  private static Input ofStream(StreamBatchReadable streamBatchReadable) {
```

Comment:
Not sure what this method is for if it is private. Just make the StreamInput ctor take all the arguments and pass that from the public ofStream()? Otherwise you will refactor this again when you remove StreamBatchReadable (which you are deprecating now, right?)